README.md
# cxl-lib

## Go Language utility library for CXL device information.

#### Copyright (c) 2023 Seagate Technology LLC and/or its Affiliates, All Rights Reserved

## Introduction

***cxl-lib*** is a standalone package for cxl device detection, information parsing and management in Linux system. User could either import this library as a module or compile the comand line version for easy interact with the CXL devices.

Source code files of the cxl-lib open source project are available to you under [The Apache-2.0 License](https://www.apache.org/licenses/LICENSE-2.0).  The
cxl-lib project repository is maintained at https://github.com/Seagate.

**cxl** Compute Express Link (CXL) is an open standard for high-speed, high capacity central processing unit (CPU)-to-device and CPU-to-memory connections, designed for high performance data center computers. 

***cxl-lib*** follows CXL specification rev 1.1.  


## Usage

```go
package main

import (
    "fmt"
    "github.com/Seagate/cxl-lib/pkg/cxl"
)

func main() {
	devList := cxl.InitCxlDevList()

    prFmt := "%12s | %20s | %10s | %10s | %15s \n"
    fmt.Printf("Print the list of CXL devs. Total devices found: %d\n", len(devList))
    fmt.Printf(prFmt, "BUS:DEV.FUN", "Vendor", "Device", "Rev", "Type")
    for _, dev := range devList {
        vendorName := dev.GetVendorInfo()
        if len(vendorName) > 15 {
            vendorName = vendorName[:15] + "..."
        }
        fmt.Printf(prFmt, dev.GetBdfString(), vendorName, dev.GetDeviceInfo(), dev.GetCxlRev(), dev.GetCxlType())
    }

}
```


CHANGELOG.md
# Change Log

## [1.0.0] - 2023/05/04
- Initial Public Release

pkg/cxl/bitfield.go
// Copyright (c) 2023 Seagate Technology LLC and/or its Affiliates

// This file implements the bitfield handling for binary structure.
// Code largely leveraged from go lang's "encoding/binary" library,
// which enables field parsing at Byte level. This file extends the
// capacity into bit level.

package cxl

import (
	"encoding/binary"
	"errors"
	"io"
	"math"
	"reflect"

	"k8s.io/klog/v2"
)

type bitfield_1b uint8
type bitfield_2b uint8
type bitfield_3b uint8
type bitfield_4b uint8
type bitfield_5b uint8
type bitfield_6b uint8
type bitfield_7b uint8
type bitfield_8b uint8
type bitfield_9b uint16
type bitfield_10b uint16
type bitfield_11b uint16
type bitfield_12b uint16
type bitfield_13b uint16
type bitfield_14b uint16
type bitfield_15b uint16
type bitfield_16b uint16
type bitfield_17b uint32
type bitfield_18b uint32
type bitfield_19b uint32
type bitfield_20b uint32
type bitfield_21b uint32
type bitfield_22b uint32
type bitfield_23b uint32
type bitfield_24b uint32
type bitfield_25b uint32
type bitfield_26b uint32
type bitfield_27b uint32
type bitfield_28b uint32
type bitfield_29b uint32
type bitfield_30b uint32
type bitfield_31b uint32
type bitfield_32b uint32
type bitfield_64b uint64

// dataSize returns the number of bytes the actual data represented by v occupies in memory.
// For compound structures, it sums the sizes of the elements. Thus, for instance, for a slice
// it returns the length of the slice times the element size and does not count the memory
// occupied by the header. If the type of v is not acceptable, dataSize returns -1.
func dataSize(v reflect.Value) int {
	klog.V(4).InfoS("bitfield.dataSize", "type", v.Type().Name(), "kind", v.Kind().String())
	switch v.Kind() {
	case reflect.Slice, reflect.Array:
		klog.V(4).InfoS("bitfield.dataSize", "len", v.Len())
		if v.Len() == 0 { // deal with empty slice
			return 0
		}
		if s := dataSize(v.Index(0)); s >= 0 {
			return s * v.Len()
		}
		return -1

	case reflect.Struct:
		sum := 0
		for i, n := 0, v.NumField(); i < n; i++ {
			s := dataSize(v.Field(i))
			if s < 0 {
				return -1
			}
			sum += s
		}
		return sum

	default:
		klog.V(2).InfoS("bitfield.dataSize", "Size", v.Type().Size())
		return int(v.Type().Size())
	}
}

// Read reads structured binary data from r into data.
// Data must be a pointer to a fixed-size value or a slice
// of fixed-size values.
// Bytes read from r are decoded using the specified byte order
// and written to successive fields of the data.
// When decoding boolean values, a zero byte is decoded as false, and
// any other non-zero byte is decoded as true.
// When reading into structs, the field data for fields with
// blank (_) field names is skipped; i.e., blank field names
// may be used for padding.
// When reading into a struct, all non-blank fields must be exported
// or Read may panic.
//
// The error is EOF only if no bytes were read.
// If an EOF happens after reading some but not all the bytes,
// Read returns ErrUnexpectedEOF.

// PCIE uses little endian
func BitFieldRead(r io.Reader, data any) error {
	order := binary.LittleEndian

	v := reflect.ValueOf(data)
	size := -1
	klog.V(4).InfoS("bitfield.BitFieldRead", "type", reflect.TypeOf(data).String(), "kind", v.Kind().String())
	switch v.Kind() {
	case reflect.Pointer:
		v = v.Elem()
		size = dataSize(v)
	case reflect.Slice:
		size = dataSize(v)
	}
	if size < 0 {
		return errors.New("bitfield.BitFieldRead: invalid type " + reflect.TypeOf(data).String())
	}
	bitLengthInfoMap := bitSizeOfArray(v)

	d := &decoder{order: order, buf: make([]byte, size)}
	ReadByBit(r, d.buf, bitLengthInfoMap)
	d.value(v)
	return nil
}

func ReadByBit(r io.Reader, buf []byte, m []int) {
	rBuf := make([]byte, 4096) // Unable to decide the length of r. Support max of 4096 bytes
	io.ReadFull(r, rBuf)

	bitOfs := 0
	i := 0
	for _, width := range m {
		endBit := bitOfs + width - 1
		startByte := bitOfs >> 3
		endByte := endBit >> 3
		bitWidthMask := uint64((1 << width) - 1)
		bitShift := bitOfs - startByte*8
		val := uint64(0)
		klog.V(4).InfoS("bitfield.ReadByBit", "bitOfs", bitOfs, "bitWidth", width, "endBit", endBit, "startByte", startByte, "endByte", endByte, "bitShift", bitShift, "bitWidthMask", hex(bitWidthMask))

		// extract related field into a uint32, and then apply the shift and mask
		structByteSize := endByte - startByte
		if structByteSize < 8 {
			for iShift := 0; iShift <= structByteSize; iShift++ {
				val |= uint64(rBuf[startByte+iShift]) << (8 * iShift)
			}
		} else {
			klog.Fatal("unsupported width!")
		}
		val = (val >> uint64(bitShift)) & bitWidthMask

		dataByteSize := (width - 1) >> 3
		if dataByteSize < 8 {
			for iShift := 0; iShift <= dataByteSize; iShift++ {
				buf[i+iShift] = byte(val >> (8 * iShift))
			}
		} else {
			klog.Fatal("unsupported width!")
		}
		i += 1 + dataByteSize
		bitOfs += width
	}
}

// sizeof returns a list of the bit offset for each field
func bitSizeOfArray(v reflect.Value) []int {

	t := v.Type()
	switch t {
	case reflect.TypeOf(bitfield_1b(0)):
		return []int{1}
	case reflect.TypeOf(bitfield_2b(0)):
		return []int{2}
	case reflect.TypeOf(bitfield_3b(0)):
		return []int{3}
	case reflect.TypeOf(bitfield_4b(0)):
		return []int{4}
	case reflect.TypeOf(bitfield_5b(0)):
		return []int{5}
	case reflect.TypeOf(bitfield_6b(0)):
		return []int{6}
	case reflect.TypeOf(bitfield_7b(0)):
		return []int{7}
	case reflect.TypeOf(bitfield_8b(0)):
		return []int{8}
	case reflect.TypeOf(bitfield_9b(0)):
		return []int{9}
	case reflect.TypeOf(bitfield_10b(0)):
		return []int{10}
	case reflect.TypeOf(bitfield_11b(0)):
		return []int{11}
	case reflect.TypeOf(bitfield_12b(0)):
		return []int{12}
	case reflect.TypeOf(bitfield_13b(0)):
		return []int{13}
	case reflect.TypeOf(bitfield_14b(0)):
		return []int{14}
	case reflect.TypeOf(bitfield_15b(0)):
		return []int{15}
	case reflect.TypeOf(bitfield_16b(0)):
		return []int{16}
	case reflect.TypeOf(bitfield_17b(0)):
		return []int{17}
	case reflect.TypeOf(bitfield_18b(0)):
		return []int{18}
	case reflect.TypeOf(bitfield_19b(0)):
		return []int{19}
	case reflect.TypeOf(bitfield_20b(0)):
		return []int{20}
	case reflect.TypeOf(bitfield_21b(0)):
		return []int{21}
	case reflect.TypeOf(bitfield_22b(0)):
		return []int{22}
	case reflect.TypeOf(bitfield_23b(0)):
		return []int{23}
	case reflect.TypeOf(bitfield_24b(0)):
		return []int{24}
	case reflect.TypeOf(bitfield_25b(0)):
		return []int{25}
	case reflect.TypeOf(bitfield_26b(0)):
		return []int{26}
	case reflect.TypeOf(bitfield_27b(0)):
		return []int{27}
	case reflect.TypeOf(bitfield_28b(0)):
		return []int{28}
	case reflect.TypeOf(bitfield_29b(0)):
		return []int{29}
	case reflect.TypeOf(bitfield_30b(0)):
		return []int{30}
	case reflect.TypeOf(bitfield_31b(0)):
		return []int{31}
	case reflect.TypeOf(bitfield_32b(0)):
		return []int{32}
	case reflect.TypeOf(bitfield_64b(0)):
		return []int{64}

	}

	switch t.Kind() {
	case reflect.Array, reflect.Slice:
		structArray := []int{}
		if v.Len() != 0 {
			s := bitSizeOfArray(v.Index(0))
			for i, n := 0, v.Len(); i < n; i++ {
				structArray = append(structArray, s[:]...)
			}
		}
		return structArray
	case reflect.Struct:
		structArray := []int{}
		for i, n := 0, t.NumField(); i < n; i++ {
			s := bitSizeOfArray(v.Field(i))
			structArray = append(structArray, s[:]...)
		}
		return structArray

	case reflect.Bool,
		reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64,
		reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
		reflect.Float32, reflect.Float64, reflect.Complex64, reflect.Complex128:
		return []int{int(t.Size()) * 8}
	}
	klog.V(2).InfoS("bitfield.bitSizeOfArray error", "kind", t.Kind().String())

	return []int{}
}

type decoder struct {
	order  binary.ByteOrder
	buf    []byte
	offset int
}

func (d *decoder) bool() bool {
	x := d.buf[d.offset]
	d.offset++
	return x != 0
}

func (d *decoder) uint8() uint8 {
	x := d.buf[d.offset]
	d.offset++
	return x
}

func (d *decoder) uint16() uint16 {
	x := d.order.Uint16(d.buf[d.offset : d.offset+2])
	d.offset += 2
	return x
}

func (d *decoder) uint32() uint32 {
	x := d.order.Uint32(d.buf[d.offset : d.offset+4])
	d.offset += 4
	return x
}

func (d *decoder) uint64() uint64 {
	x := d.order.Uint64(d.buf[d.offset : d.offset+8])
	d.offset += 8
	return x
}

func (d *decoder) int8() int8 { return int8(d.uint8()) }

func (d *decoder) int16() int16 { return int16(d.uint16()) }

func (d *decoder) int32() int32 { return int32(d.uint32()) }

func (d *decoder) int64() int64 { return int64(d.uint64()) }

func (d *decoder) value(v reflect.Value) {
	switch v.Kind() {
	case reflect.Array:
		l := v.Len()
		for i := 0; i < l; i++ {
			d.value(v.Index(i))
		}

	case reflect.Struct:
		t := v.Type()
		l := v.NumField()
		for i := 0; i < l; i++ {
			// Note: Calling v.CanSet() below is an optimization.
			// It would be sufficient to check the field name,
			// but creating the StructField info for each field is
			// costly (run "go test -bench=ReadStruct" and compare
			// results when making changes to this code).
			if v := v.Field(i); v.CanSet() || t.Field(i).Name != "_" {
				d.value(v)
			} else {
				d.skip(v)
			}
		}

	case reflect.Slice:
		l := v.Len()
		for i := 0; i < l; i++ {
			d.value(v.Index(i))
		}

	case reflect.Bool:
		v.SetBool(d.bool())

	case reflect.Int8:
		v.SetInt(int64(d.int8()))
	case reflect.Int16:
		v.SetInt(int64(d.int16()))
	case reflect.Int32:
		v.SetInt(int64(d.int32()))
	case reflect.Int64:
		v.SetInt(d.int64())

	case reflect.Uint8:
		v.SetUint(uint64(d.uint8()))
	case reflect.Uint16:
		v.SetUint(uint64(d.uint16()))
	case reflect.Uint32:
		v.SetUint(uint64(d.uint32()))
	case reflect.Uint64:
		v.SetUint(d.uint64())

	case reflect.Float32:
		v.SetFloat(float64(math.Float32frombits(d.uint32())))
	case reflect.Float64:
		v.SetFloat(math.Float64frombits(d.uint64()))

	case reflect.Complex64:
		v.SetComplex(complex(
			float64(math.Float32frombits(d.uint32())),
			float64(math.Float32frombits(d.uint32())),
		))
	case reflect.Complex128:
		v.SetComplex(complex(
			math.Float64frombits(d.uint64()),
			math.Float64frombits(d.uint64()),
		))
	}
}

func (d *decoder) skip(v reflect.Value) {
	d.offset += dataSize(v)
}


pkg/cxl/cxl-util.go
// Copyright (c) 2023 Seagate Technology LLC and/or its Affiliates

// This file implements the API functions of the cxl library
package cxl

import (
	"bufio"
	"bytes"
	"fmt"
	"os"
	"strconv"
	"strings"
	"syscall"

	"github.com/jaypipes/pcidb"
	"k8s.io/klog/v2"
)

// Base address of PCI memory mapped configurations
var PCI_MMCONFIG_BASE_ADDR int64

// CxlDevType : The type of the CXL device, ie type 1, 2, 3.
type CxlDevType string

// List of CXL device types
const (
	CXL_UNKOWN_DEV CxlDevType = "CXLDeviceTypeUnkown"
	CXL_TYPE1_DEV  CxlDevType = "CXLType1Device"
	CXL_TYPE2_DEV  CxlDevType = "CXLType2Device"
	CXL_TYPE3_DEV  CxlDevType = "CXLType3Device"
)

// CxlRev : The revision the CXL device, ie rev 1.1, 2.0, 3.0.
type CxlRev string

// List of CXL revisions
const (
	CXL_REV_UNKOWN CxlRev = "CXL_unkown"
	CXL_REV_1_1    CxlRev = "CXL1.1"
	CXL_REV_2_0    CxlRev = "CXL2.0"
	CXL_REV_3_0    CxlRev = "CXL3.0"
)

// Capability struct for CXL device
type CxlCaps struct {
	Cache_Cap bool
	IO_Cap    bool
	Mem_Cap   bool
	Cache_En  bool
	IO_En     bool
	Mem_En    bool
}

// Memory Attribute table for CXL memory perforrmance
type CxlMemAttr struct {
	ReadLatency    uint16
	WriteLatency   uint16
	ReadBandwidth  uint16
	WriteBandwidth uint16
}

type ACPI struct {
	CEDT *cedt_table_struct
}

// ACPI tables are static, initialize via init() func
var ACPITables = ACPI{}

func init() {
	getPciMmConfig()
	ACPITables.FetchCedt()
}

// Update local copy of the cedt .
func (a *ACPI) FetchCedt() {
	b, err := readACPI("CEDT")
	if err == nil {
		acpiHdr := parseStruct(b, ACPI_HEADER{})
		cedtFileS := parseStruct(b, CEDT_TABLE(uint(acpiHdr.getCedtSubtableCountFromAcpiHeader())))
		a.CEDT = &cedtFileS
	} else {
		klog.V(1).Info(err)
	}
}

// Get cedt subtable counts.
func (a *ACPI) GetCedtCount() int {
	return len(a.CEDT.Subtable)
}

// Get subtable cedt struct by index.
func (a *ACPI) GetCedtSubtable(i int) *CEDT_SUBTABLE {
	if a.CEDT == nil {
		a.FetchCedt()
	}
	if len(a.CEDT.Subtable) > i {
		return &a.CEDT.Subtable[i]
	}
	return nil
}

type CxlDev struct {
	Bdf        *BDF       `json:"BDF"`
	Vendor     string     `json:"Vendor"`
	CXLrev     CxlRev     `json:"CXL-Rev"`
	CXLdevtype CxlDevType `json:"CXL-Type"`
	PCIE       []byte     `json:"-"`
}

// initialize the structure based on BDF value
func (c *CxlDev) init(b *BDF) error {
	var err error = nil
	if b == nil {
		err = fmt.Errorf("bdf is empty")
	} else {
		c.Bdf = b
		c.updatePcieConfig()
		c.Vendor = c.GetVendorInfo()
		c.CXLrev = c.GetCxlRev()
		c.CXLdevtype = c.GetCxlType()
	}

	return err
}

// check if a device is CXL device.
func (c *CxlDev) isCxlDev() bool {
	pcieHeader := parseStruct(c.PCIE, PCIE_CONFIG_HDR{})
	klog.V(3).InfoS("InfoS structured:   cxl-util: isCXLDev", "Vendor", hex(pcieHeader.Vendor_ID), "device", hex(pcieHeader.Device_ID), "class", hex(pcieHeader.Class_Code.Base_Class_Code), "sub", hex(pcieHeader.Class_Code.Sub_Class_Code), "prog-if", hex(pcieHeader.Class_Code.Prog_if))
	if pcieHeader.Class_Code.Base_Class_Code == 0x5 && // 0x05: Memory Controller
		pcieHeader.Class_Code.Sub_Class_Code == 0x2 && // 0x02: CXL memory devic
		pcieHeader.Class_Code.Prog_if == 0x10 { // 0x10: Always 0x10 per spec
		return true
	}
	return false
}

// Update local copy of the pcie config .
func (c *CxlDev) updatePcieConfig() {
	c.PCIE = readMemory4k(c.Bdf.bdfToMemAddr())
}

// return the BDF as string BUS:DEV.FUN
func (c *CxlDev) GetBdfString() string {
	return fmt.Sprintf("%02X:%02X.%1X", c.Bdf.Bus, c.Bdf.Device, c.Bdf.Function)
}

// return the value of the CEDT field
func (c *CxlDev) GetCedtField(field string) interface{} {
	return nil
}

// return a list of DVSEC tables from the CXL device
func (c *CxlDev) GetDvsecList() map[cxl_dvsec_id]uint32 {
	dvsecMap := make(map[cxl_dvsec_id]uint32)
	next_cap := uint32(EXT_DVSEC_OFFSET)
	for next_cap != 0 {
		pcieCapHeader := parseStruct(c.PCIE[next_cap:], PCIE_EXT_CAP_HDR{})
		if int(pcieCapHeader.DVSEC_hdr1.DVSEC_Vendor_ID) == CXL_Vendor_ID {
			dvsec_id := pcieCapHeader.DVSEC_hdr2.DVSEC_ID
			dvsecMap[cxl_dvsec_id(dvsec_id)] = next_cap
		}
		next_cap = uint32(pcieCapHeader.Next_Cap_ofs)
	}
	return dvsecMap
}

// return the struct of a DVSEC
func (c *CxlDev) GetDvsec(dvsecId cxl_dvsec_id) interface{} {
	Dvseclist := c.GetDvsecList()
	Dvsecoffset, ok := Dvseclist[dvsecId]
	if !ok {
		klog.Error("can't find Dvseclist")
	}
	switch dvsecId {
	case CXL_DVSEC_PCIE_DVSEC_FOR_CXL:
		return parseStruct(c.PCIE[Dvsecoffset:], PCIE_DVSEC_FOR_CXL{})
	case CXL_DVSEC_NON_CXL_FUNC_MAP:
		return parseStruct(c.PCIE[Dvsecoffset:], NON_CXL_FUNC_MAP{})
	case CXL_DVSEC_CXL2_0_EXR_DVESC:
		return parseStruct(c.PCIE[Dvsecoffset:], CXL2_0_EXR_DVESC{})
	case CXL_DVSEC_GPF_DVSEC_FOR_PORTS:
		return parseStruct(c.PCIE[Dvsecoffset:], GPF_DVSEC_FOR_PORTS{})
	case CXL_DVSEC_GPF_DVSEC_FOR_DEV:
		return parseStruct(c.PCIE[Dvsecoffset:], GPF_DVSEC_FOR_DEV{})
	case CXL_DVSEC_PCIE_DVSEC_FOR_FLEX_BUS_PORT:
		return parseStruct(c.PCIE[Dvsecoffset:], PCIE_DVSEC_FOR_FLEX_BUS_PORT{})
	case CXL_DVSEC_REGISTER_LOCATOR:
		RegisterLocator := parseStruct(c.PCIE[Dvsecoffset:], registerLocator{})
		BlockNumber := RegisterLocator.getRegisterBlockNumberFromHeader()
		return parseStruct(c.PCIE[Dvsecoffset:], REGISTER_LOCATOR(BlockNumber))
	case CXL_DVSEC_MLD:
		return parseStruct(c.PCIE[Dvsecoffset:], MLD{})
	case CXL_DVSEC_PCIE_DVSEC_FOR_TEST_CAP:
		return parseStruct(c.PCIE[Dvsecoffset:], PCIE_DVSEC_FOR_TEST_CAP{})
	}
	return nil
}

// return the CXL revision
func (c *CxlDev) GetCxlRev() CxlRev {
	Dvseclist := c.GetDvsecList()
	_, ok := Dvseclist[7]
	if !ok {
		return CXL_REV_1_1
	} else {
		CXLdvsec := c.GetDvsec(CXL_DVSEC_PCIE_DVSEC_FOR_FLEX_BUS_PORT).(PCIE_DVSEC_FOR_FLEX_BUS_PORT)
		switch CXLdvsec.PCIE_ext_cap_hdr.DVSEC_hdr1.DVSEC_Rev {
		case 0:
			return CXL_REV_1_1
		case 1:
			return CXL_REV_2_0
		case 2:
			return CXL_REV_3_0
		default:
			return CXL_REV_UNKOWN
		}
	}

}

// return the capacities info of the CXL device
func (c *CxlDev) GetCxlCap() CxlCaps {
	Dvsecforcxl := c.GetDvsec(CXL_DVSEC_PCIE_DVSEC_FOR_CXL).(PCIE_DVSEC_FOR_CXL)
	return CxlCaps{
		Cache_Cap: UintToBool(Dvsecforcxl.CXL_cap.Cache_Cap),
		IO_Cap:    UintToBool(Dvsecforcxl.CXL_cap.IO_Cap),
		Mem_Cap:   UintToBool(Dvsecforcxl.CXL_cap.Mem_Cap),
		Cache_En:  UintToBool(Dvsecforcxl.CXL_ctrl.Cache_En),
		IO_En:     UintToBool(Dvsecforcxl.CXL_ctrl.IO_En),
		Mem_En:    UintToBool(Dvsecforcxl.CXL_ctrl.Mem_En),
	}
}

// convert integer to bool
func UintToBool(i bitfield_1b) bool {
	if i == 1 {
		return true
	} else {
		return false
	}
}

// return the type info of the CXL device ( type 1/ type 2/ type 3 )
// Type 1 - CXL.cache and CXL.io
// Type 2 - CXM.mem and CXL.cache and CXL.io
// Type 3 - CXL.mem and CXL.io
func (c *CxlDev) GetCxlType() CxlDevType {
	CXLCAP := c.GetCxlCap()
	if CXLCAP.Cache_En && CXLCAP.IO_En && !CXLCAP.Mem_En {
		return CXL_TYPE1_DEV
	} else if CXLCAP.Mem_En && CXLCAP.Cache_En && CXLCAP.IO_En {
		return CXL_TYPE2_DEV
	} else if CXLCAP.Mem_En && CXLCAP.IO_En && !CXLCAP.Cache_En {
		return CXL_TYPE3_DEV
	}
	return CXL_UNKOWN_DEV
}

// return the memory size of a CXL device
// Memory_Size_High: Corresponds to bits 63:32 of the CXL Range 1 memory size regardless of whether the device implements CXL HDM Decoder Capability registers.
// Memory_Size_Low: Corresponds to bits 31:28 of the CXL Range 1 memory size regardless of whether the device implements CXL HDM Decoder Capability registers.
func (c *CxlDev) GetMemorySize() int64 {
	dvsecForCxl := c.GetDvsec(CXL_DVSEC_PCIE_DVSEC_FOR_CXL).(PCIE_DVSEC_FOR_CXL)
	return (int64(dvsecForCxl.CXL_range1_size_high.Memory_Size_High) << 32) | (int64(dvsecForCxl.CXL_range1_size_low.Memory_Size_low) << 28)
}

// return the memory base of a CXL device
func (c *CxlDev) GetMemoryBaseAddr() int64 {
	dvsecForCxl := c.GetDvsec(CXL_DVSEC_PCIE_DVSEC_FOR_CXL).(PCIE_DVSEC_FOR_CXL)
	return (int64(dvsecForCxl.CXL_range1_base_high.Memory_Base_High) << 32) | (int64(dvsecForCxl.CXL_range1_base_low.Memory_Base_Low) << 28)
}

// return the pcie header struct
func (c *CxlDev) GetPcieHdr() *PCIE_CONFIG_HDR {
	pcieHeader := parseStruct(c.PCIE, PCIE_CONFIG_HDR{})
	return &pcieHeader
}

// return the Vendor Info of the PCIe/CXL device
func (c *CxlDev) GetVendorInfo() string {
	//map of vendor
	Pcidbs, err := pcidb.New()
	if err != nil {
		fmt.Printf("Error getting PCI info: %v", err)
	}
	pcieHeader := parseStruct(c.PCIE, PCIE_CONFIG_HDR{})
	return Pcidbs.Vendors[fmt.Sprintf("%x", pcieHeader.Vendor_ID)].Name
}

// return the Vendor Info of the PCIe/CXL device
func (c *CxlDev) GetDeviceInfo() string {
	pcieHeader := parseStruct(c.PCIE, PCIE_CONFIG_HDR{})
	return fmt.Sprintf("0x%X", pcieHeader.Device_ID)
}

// obtain a list of CXL devices on the host
func InitCxlDevList() map[string]*CxlDev {
	CxlDevMap := make(map[string]*CxlDev)

	pcieDevPath := "/sys/bus/pci/devices"
	links, err := os.ReadDir(pcieDevPath)
	if err != nil {
		klog.Fatal(err)
	}
	for _, link := range links {
		// init BDF struct
		bdf := BDF{}
		// Convert the Linux fs format to structure
		bdf.addrToBDF(link.Name())
		klog.V(3).InfoS("cxl-util.InitCxlDevList", "Addr", hex(bdf.bdfToMemAddr()))
		if checkCxlDevClass(link.Name()) {
			new_CxlDev := CxlDev{}
			err = new_CxlDev.init(&bdf)
			if err == nil && new_CxlDev.isCxlDev() {
				klog.V(2).InfoS("cxl-util.InitCxlDevList Device found", "Link", link.Name())
				CxlDevMap[new_CxlDev.GetBdfString()] = &new_CxlDev
			}
		}

	}
	return CxlDevMap
}

func checkCxlDevClass(link string) bool {
	path := fmt.Sprintf("/sys/bus/pci/devices/%s/class", link)
	fileBytes, err := os.ReadFile(path)
	klog.V(3).InfoS("cxl-util.checkCxlDevClass", "Link", path, "file", fileBytes)
	if fileBytes != nil && err == nil {
		if string(fileBytes) == "0x050210\n" {
			return true
		}
	}

	return false
}

// readMemory4k: return a 4k sized byte array from the memory physical address
func readMemory4k(baseAddress int64) []byte {
	const bufferSize int = 4096

	// Check for 4k boundary align
	klog.V(2).InfoS("cxl-util.readMemory4k", "BaseAddress", hex(baseAddress))
	if baseAddress&int64(bufferSize-1) != 0 {
		klog.Fatal(fmt.Errorf("BaseAddress is not 4k aligned"))
	}

	file, err := os.Open("/dev/mem")
	if err != nil {
		klog.Fatal(err)
	}
	klog.V(3).Info("cxl-util.readMemory4k /dev/mem is opened")

	defer file.Close()

	mmap, err := syscall.Mmap(int(file.Fd()), baseAddress, bufferSize, syscall.PROT_READ, syscall.MAP_SHARED)
	if err != nil {
		klog.Fatal(err)
	}
	klog.V(3).Info("cxl-util.readMemory4k Mmap is done")

	mmapCp := make([]byte, bufferSize)
	// Save a copy of mmap, which will be elimicated after syscall.Munmap(mmap)
	for i := 0; i < bufferSize; i++ {
		mmapCp[i] = mmap[i]
	}
	err = syscall.Munmap(mmap)
	if err != nil {
		klog.Fatal(err)
	}
	klog.V(3).Info("cxl-util.readMemory4k Munmap is done")
	return mmapCp
}

// readACPI: return the byte array of an ACPI table matching the input string
func readACPI(t string) ([]byte, error) {
	var path string = ""
	var err error
	var fileBytes []byte
	// Check for ACPI table name
	if t == "" {
		err = fmt.Errorf("error input: ACPI table name is not supplied")
	} else {
		path = fmt.Sprintf("/sys/firmware/acpi/tables/%s", strings.ToUpper(t))
		fileBytes, err = os.ReadFile(path)
		if err == nil {
			if len(fileBytes) != 0 {
				hdr := parseStruct(fileBytes, ACPI_HEADER{})
				if fmt.Sprintf("%s", hdr.Signature) == t {
					return fileBytes, nil
				} else {
					err = fmt.Errorf("error signature: file signature doesn't match")
				}
			} else {
				err = fmt.Errorf("error read: file read return 0 bytes")
			}
		}
	}
	return nil, err
}

// getPciMmConfig: return the PCI_MMCONFIG value from /proc/iomem
func getPciMmConfig() {
	readFile, err := os.Open("/proc/iomem")
	if err != nil {
		klog.Fatal(err)
	}
	defer readFile.Close()

	fileScanner := bufio.NewScanner(readFile)
	fileScanner.Split(bufio.ScanLines)
	for fileScanner.Scan() {
		text := fileScanner.Text()
		if strings.Contains(text, "MMCONFIG") {
			// String Example "   80000000-8fffffff : PCI MMCONFIG 0000 [bus 00-ff]"
			PCI_MMCONFIG_BASE_ADDR = int64(hexToInt(strings.TrimSpace(strings.Split(fileScanner.Text(), "-")[0])))
			break
		}
	}
}

// parse binary array into struct.
func parseStruct[T any](b []byte, s T) T {
	buf := &bytes.Buffer{}
	buf.Write(b)
	newStruct := s
	err := BitFieldRead(buf, &newStruct)
	if err != nil {
		klog.Fatal(err)
	}
	return newStruct
}

type BDF struct {
	Domain   uint16 `json:"Domain"`
	Bus      uint8  `json:"Bus"`
	Device   uint8  `json:"Device"`
	Function uint8  `json:"Function"`
}

func (b *BDF) addrToBDF(addr string) {
	bdfStringList := strings.Split(strings.ToLower(addr), ":")
	if len(bdfStringList) != 3 {
		klog.Fatal(fmt.Errorf("address format error. Expect $domain:$bus:$dev.$func"))
	}
	dfStringList := strings.Split(bdfStringList[2], ".")
	if len(dfStringList) != 2 {
		klog.Fatal(fmt.Errorf("address format error. Expect $domain:$bus:$dev.$func"))
	}

	b.Domain = uint16(hexToInt(bdfStringList[0]))
	b.Bus = uint8(hexToInt(bdfStringList[1]))
	b.Device = uint8(hexToInt(dfStringList[0]))
	b.Function = uint8(hexToInt(dfStringList[1]))
}

func (b *BDF) bdfToMemAddr() int64 {
	return PCI_MMCONFIG_BASE_ADDR | (int64(b.Function) << 12) | (int64(b.Device) << 15) | (int64(b.Bus) << 20)
}

func hexToInt(hexStr string) uint64 {
	// base 16 for hexadecimal
	result, _ := strconv.ParseUint(hexStr, 16, 64)
	return result
}

// Wrapper function to shorten int to hex convertion call
func hex(a any) string {
	return fmt.Sprintf("%X", a)
}
